File: back-end/src/main/java/RealEstate.java
package main.java;

import java.io.IOException;
import java.util.ArrayList;

public abstract class RealEstate extends User {

    protected String uri;
    protected long lastTimestamp;

    public RealEstate(String username, String uri) {
        super(username);
        this.uri = uri;
        this.lastTimestamp = 0;
    }

    public abstract ArrayList<House> getHouses() throws IOException;

    public abstract House getHouse(String id) throws IOException;

    public long getLastTimestamp() {
        return lastTimestamp;
    }
}

File: back-end-prototype/models/realEstate.js
'use strict';
const User = require('./user');
var debug = require('debug')('khanebedoosh:models');

class RealEstate extends User {
    constructor(username, uri, getHouse, getHouses) {
        super(username);
        this._uri = uri;
        this._lastTimestamp = 0;
        this.getHouse = getHouse;
        this.getHouses = async () => {
            var { houses, expireTime } = await getHouses();
            this._lastTimestamp = expireTime;
            return houses;
        };
    }

    get uri() {
        return this._uri;
    }

    get lastTimestamp() {
        return this._lastTimestamp;
    }

    set lastTimestamp(lastTimestamp) {
        this._lastTimestamp = lastTimestamp;
    }

    get json() {
        var res = {};
        res.uri = this.uri;
        res.lastTimestamp = this.lastTimestamp;
        return { .  ..super.json, ...res };
    }
}

module.exports = RealEstate;
